encoding/protobuf/observability_test.go

```diff
@@ -0,0 +1,221 @@
+	require.Len(t, counterAssertions, len(snapshot), "unexpected number of counters")
+
+	for i, wantCounter := range counterAssertions {
+		require.Equal(t, wantCounter.Name, snapshot[i].Name, "unexpected counter")
+		assert.EqualValues(t, wantCounter.Value, snapshot[i].Value, "unexpected counter value")
+		for wantTagKey, wantTagVal := range wantCounter.Tags {
+			assert.Equal(t, wantTagVal, snapshot[i].Tags[wantTagKey], "unexpected value for %q", wantTagKey)
+		}
+	}
+}
+
+func initClientAndServer(t *testing.T) (
+	client testpb.TestYARPCClient,
+	observedLogs *observer.ObservedLogs,
+	clientMetricsRoot *metrics.Root,
+	serverMetricsRoot *metrics.Root,
+	cleanup func(),
+) {
+	loggerCore, observedLogs := observer.New(zapcore.DebugLevel)
+	clientMetricsRoot, serverMetricsRoot = metrics.New(), metrics.New()
+
+	serverAddr, cleanupServer := newServer(t, loggerCore, serverMetricsRoot)
+	client, cleanupClient := newClient(t, serverAddr, loggerCore, clientMetricsRoot)
+
+	_ = observedLogs.TakeAll() // ignore all start up logs
+
+	return client, observedLogs, clientMetricsRoot, serverMetricsRoot, func() {
+		cleanupServer()
+		cleanupClient()
+	}
+}
+
+func newServer(t *testing.T, loggerCore zapcore.Core, metricsRoot *metrics.Root) (addr string, cleanup func()) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+
+	inbound := grpc.NewTransport().NewInbound(listener)
+	dispatcher := yarpc.NewDispatcher(yarpc.Config{
+		Name:     _serverName,
+		Inbounds: yarpc.Inbounds{inbound},
+		Logging:  yarpc.LoggingConfig{Zap: zap.New(loggerCore)},
+		Metrics:  yarpc.MetricsConfig{Metrics: metricsRoot.Scope()},
+	})
+
+	dispatcher.Register(testpb.BuildTestYARPCProcedures(&observabilityTestServer{}))
+	require.NoError(t, dispatcher.Start(), "could not start server dispatcher")
+
+	addr = inbound.Addr().String()
+	cleanup = func() { assert.NoError(t, dispatcher.Stop(), "could not stop dispatcher") }
+	return addr, cleanup
+}
+
+func newClient(t *testing.T, serverAddr string, loggerCore zapcore.Core, metricsRoot *metrics.Root) (client testpb.TestYARPCClient, cleanup func()) {
+	dispatcher := yarpc.NewDispatcher(yarpc.Config{
+		Name: _clientName,
+		Outbounds: map[string]transport.Outbounds{
+			_serverName: {
+				ServiceName: _serverName,
+				Unary:       grpc.NewTransport().NewSingleOutbound(serverAddr),
+			},
+		},
+		Logging: yarpc.LoggingConfig{Zap: zap.New(loggerCore)},
+		Metrics: yarpc.MetricsConfig{Metrics: metricsRoot.Scope()},
+	})
+
+	client = testpb.NewTestYARPCClient(dispatcher.ClientConfig(_serverName))
+	require.NoError(t, dispatcher.Start(), "could not start client dispatcher")
+
+	cleanup = func() { assert.NoError(t, dispatcher.Stop(), "could not stop dispatcher") }
+	return client, cleanup
+}
+
+type observabilityTestServer struct{}
+
+func (observabilityTestServer) Unary(context.Context, *testpb.TestMessage) (*testpb.TestMessage, error) {
+	deets := []proto.Message{
```

Comment:
optional: maybe `deets` is a bit too colloquial 😄  maybe change to `details`, up to you though

Reply:
sometimes I have too much fun, I'll change it :D